### PR TITLE
pipes - anders - api muncher

### PR DESCRIPTION
**Pull request by anderschenders**
# API Muncher
Congratulations! You're submitting your assignment!

## Comprehension Questions
| Question | Answer    |
| :------------- | :------------- |
| How did you go about exploring the Edamam API, how did you try querying the API? | I read the Edamam API documentation. I used Postman to see what data would be sent back and in what format, depending on different queries. |
| Describe your API Wrapper.  How did you decide on the methods you created? | I created 2 methods, one for a search term and one to find a specific recipe. I decided to split these into 2 methods as they were two separate actions. I also created 2 private methods, one to help DRY up my code and the other to help create objects with the relevant data from the API request. |
| Describe an edge case or failure case test you wrote for your API Wrapper. | I tested for the edge case where a bad id and key were passed in along with a valid search term. I was testing that an error would be raised in this case. |
| Explain how VCR aids in testing an API. | API requests take time, can be costly and depend on the API to be functioning well, and since we need to make calls to the API for many tests, VCR helps this process by recording a call to the API and allowing us to reuse it for multiple tests. |
| What is the Heroku URL of your deployed application? | https://pacific-fjord-78739.herokuapp.com/ |

### CODE REVIEW

File: app/views/layouts/application.html.erb [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+          <%= label_tag(:q, "Search Recipes:") %>
+          <%= text_field_tag(:q) %>
+          <%= submit_tag("Search", class:"button")%>
+        <% end %>
+      </div>
+    </nav>
+  </header>
+  <% end %>
+
+  <body>
```
Comment (droberts-sea) [inline app/views/layouts/application.html.erb]: Your `<header>` should be inside the `<body>`.

File: app/controllers/recipes_controller.rb [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+    end
+
+  end
+
+  def show
+    # uri = params[:id]
+    # uri.gsub(/%2F/, "/")
+    begin
+      @search = params[:search_term] #not working - how do I pass this param in properly?
+      @recipe = EdamamApiWrapper.find_recipe(params[:id])
```
Comment (droberts-sea) [inline app/controllers/recipes_controller.rb:37]: I don't understand this comment. I don't believe that your show page should have a `search_term` in the query params.

File: lib/edamam_api_wrapper.rb [hunk 3]
```diff
@@ -0,0 +1,71 @@
+require 'httparty'
+
+class EdamamApiWrapper
+  BASE_URL = "https://api.edamam.com/"
+  BASE_URI = "?r=http://www.edamam.com/ontologies/edamam.owl%23recipe_"
```
Comment (droberts-sea) [inline lib/edamam_api_wrapper.rb:5]: Calling these `BASE_URL` and `BASE_URI` is a bit confusing - I would probably have called the second one something like `RECIPE_URI_PREFIX`.

File: test/controllers/recipes_controller_test.rb [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+      VCR.use_cassette("norecipes") do
+        q = {
+          q: ""
+        }
+
+        get recipes_path, params: q
+        flash[:status].must_equal :failure
+        must_respond_with :redirect
+      end
+    end
```
Comment (droberts-sea) [inline test/controllers/recipes_controller_test.rb:57]: I like these error cases around the search term. You should have some testing here around the pagination parameters. In particular:

- Test that using paging params gives different results (or at least that it works)
- What if the paging parameters are invalid (negative number, etc)?
- What if the paging parameters take you past the end of the available search results?

File: test/controllers/recipes_controller_test.rb [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+        id = recipes[0].uri_id
+
+        get recipe_path(id)
+        must_respond_with :success
+      end
+    end
+
+    it "should set flash[:status] to :failure and redirect_back if given invalid data" do
+      VCR.use_cassette("no_recipe") do
+      	get recipe_path("Not a valid uri")
```
Comment (droberts-sea) [inline test/controllers/recipes_controller_test.rb:110]: Good!

File: .gitignore [hunk 6]
```diff
@@ -0,0 +1,23 @@
...
+!/tmp/.keep
+
+/node_modules
+/yarn-error.log
+
+.byebug_history
+
+# ignore application id and key
+.env
+/test/cassettes/*
```
Comment (droberts-sea) [inline .gitignore:23]: Since cassettes are part of your tests, you should include them in git.

File: test/lib/edamam_api_wrapper_test.rb [hunk 7]
```diff
@@ -0,0 +1,47 @@
...
+        end
+      end
+    end
+
+    it "should raise an ApiError when the app_id and app_key are invalid" do
+      VCR.use_cassette("apple") do
+        proc {
+          EdamamApiWrapper.search("apple", "bad_app_id", "bad_app_key")
+        }.must_raise EdamamApiWrapper::ApiError
+      end
```
Comment (droberts-sea) [inline test/lib/edamam_api_wrapper_test.rb:22]: What if the search term yields 0 results? Something like "fkljadlfjaljfaljfalkdjf" would probably do.

**Comment by droberts-sea** [comment]
# API Muncher
## What We're Looking For

| Feature | Feedback    |
| :------------- | :------------- |
| **Core Requirements** |   |
| Git hygiene | yes |
| Comprehension questions	| yes |
| **General** |  |
| Rails fundamentals (RESTful routing, use of named paths) | yes |
| Semantic HTML | yes |
| Errors are reported to the user | yes |
| API Wrapper to handle the API requests | yes |
| Controller testing | yes |
| Lib testing | yes |
| Search Functionality | yes |
| List Functionality | yes |
| Show individual item functionality (link to original recipe opens in new tab) | mostly - no link to the original recipe |
| **Styling** |  |
| Foundation Styling for responsive layout | yes |
| List View shows 10 items at a time/pagination | yes |
| The app is styled to create an attractive user interface | yes |
| **API Features** | |
| The App attributes Edaman | yes |
| The VCR casettes do not contain the API key | yes |
| **External Resources** | |
| Link to deployed app on Heroku | yes |
|  **Overall** | Good job overall! You've definitely hit the core learning goals here - the site is attractive and easy to use, it pulls data from the API, and it is able to handle errors and report them to the user politely. As with most software there are some spots where test coverage could be expanded, but this too looks pretty good. Keep up the hard work! |
